packages/frontend/app/services/channels/-utils/channel-saver.ts

```diff
@@ -0,0 +1,75 @@
+import Channel from 'emberclear/models/channel';
+import Vote from 'emberclear/models/vote';
+import VoteChain from 'emberclear/models/vote-chain';
+import ChannelContextChain from 'emberclear/models/channel-context-chain';
+import Identity from 'emberclear/models/identity';
+
+export async function saveChannel(channel: Channel) {
```

Comment:
given all this asynchrony, I wonder if all of this data needs to be in so many models, saved individually? or just in a single POJO attribute on the `Channel`? would that make things easier (if everything was just on the Channel?)

Perhaps (attempted to be described as types):
Also with questions, because I copied from the models directory, and I'm not 100% sure what each thing is
```ts
type Vote = {
  remaining: string[]; // user ids
  yes: string[]; // user ids;
  no: string[]; // user ids;
  target: Identity; // what's a target? 
  action: VOTE_ACTION;
  key: Identity; // public key of the voter
  previousVoteChain?: VoteChain;
  signature!: Uint8Array;
}
type VoteChain = {
  yes: Vote[];
  no: Vote[];
  remaining: Vote[];
  target: unknown; // what's a target?
  key: unknown; // what's a key? / which / who's key?
  previousChain: VoteChain;
}
// looking at just the two types above? is there any redundant information? 
// I'm not sure, but just by looking at the properties, it might kinda look like it?
// idk. I could have converted from the models wrong, too

class Channel extends Model {
  // ...existing stuff

  @attr members: string[] = []; // ids of the members
  @attr admin!: string; // id;
  @attr supportingVote?: VoteChain; // semantically, what does "Supporting Vote" mean?
  @attr previousChain?: VoteChain;
}
```

Reply:
I think with the new models, this may be fixed. Because of how all the relations work according to the ember docs, when you update your objects I believe it updates the other side of the relationship as well. I can step you through the logic behind the models on a call if you'd like, but I do think that we likely need all of them except for maybe Vote??? A lot of what we did in the model structuring is because it needs to be that way to be able to verify that contexts haven't been tampered with.

Reply:
The Vote and VoteChain above are redundant. We use VoteChain in the system to store actual vote data and Vote is essentially used as a wrapper such that one "vote" has a constant id registered with the channel even if the chain inside is changing.

As for what target and key are for, target is the identity of the target of the VOTE_ACTION. For example, if VOTE_ACTION was add and target was NullVoxPopuli, that vote-chain would be a vote to add NullVoxPopuli to a given channel. Same idea with remove and promote. The key of a votechain is the identity of the user who voted and signed that chain. This gives us the ability to check votechains still after they are initially sent since we otherwise wouldn't have the sender information. It is used to unsign the signature for verification as well as check the the change in yes/no/remaining from the previous vote is valid.